pkg/controller/cronjob/utils.go

```diff
@@ -121,33 +121,87 @@ func getRecentUnmetScheduleTimes(sj batchv1beta1.CronJob, now time.Time) ([]time
 		return []time.Time{}, nil
 	}
 
-	for t := sched.Next(earliestTime); !t.After(now); t = sched.Next(t) {
-		starts = append(starts, t)
-		// An object might miss several starts. For example, if
-		// controller gets wedged on friday at 5:01pm when everyone has
-		// gone home, and someone comes in on tuesday AM and discovers
-		// the problem and restarts the controller, then all the hourly
-		// jobs, more than 80 of them for one hourly scheduledJob, should
-		// all start running with no further intervention (if the scheduledJob
-		// allows concurrency and late starts).
-		//
-		// However, if there is a bug somewhere, or incorrect clock
-		// on controller's server or apiservers (for setting creationTimestamp)
-		// then there could be so many missed start times (it could be off
-		// by decades or more), that it would eat up all the CPU and memory
-		// of this controller. In that case, we want to not try to list
-		// all the missed start times.
-		//
-		// I've somewhat arbitrarily picked 100, as more than 80,
-		// but less than "lots".
-		if len(starts) > 100 {
-			// We can't get the most recent times so just return an empty slice
-			return []time.Time{}, fmt.Errorf("Too many missed start time (> 100). Set or decrease .spec.startingDeadlineSeconds or check clock skew.")
-		}
+	// For the sake of easy patching, try to "hide" getLatestMissedSchedule
+	latestTime, missedTimes := getLatestMissedSchedule(earliestTime, now, sched)
+	for i := 0; i < missedTimes; i++ {
+		starts = append(starts, latestTime) // This technically breaks the function signature, but the actual consumer only cares about starts[0] and len(starts).
 	}
+
 	return starts, nil
 }
 
+// getLatestMissedSchedule returns the latest start time in the time window (inclusive), and the number of schedules in the time window (capped at 2).
+// The number of missed start times conveys "none, one, multiple".
+func getLatestMissedSchedule(startWindow time.Time, endWindow time.Time, schedule cron.Schedule) (time.Time, int) {
+	nextSchedule := schedule.Next(startWindow)
+
+	// If no schedules in window, return.
+	if nextSchedule.After(endWindow) {
+		return time.Time{}, 0
+
+		// There is (at least one) schedule in the window. See if there are more later ones.
+	} else {
+		latestSchedule, found := getLatestMissedScheduleBinarySearch(nextSchedule, endWindow, schedule)
+		if found && latestSchedule != nextSchedule {
+			return latestSchedule, 2 // We missed at least 2 schedules. Return the latest.
+		} else {
+			return nextSchedule, 1
+		}
+	}
+}
+
+// LatestScheduleTime, foundSchedule
+// getLatestMissedSchedule returns the latest start time in the specified time window (inclusive), and a true/false indicating if a start time was found.
```

Comment:
This suggestion is just to remove any possible double-negative confusion :)
```suggestion
// getLatestMissedSchedule returns the latest start time in the specified time window (inclusive). It also returns true if a start time was found, false otherwise
```